[feat] User profileImage webp 확장자 변환 후 S3 업로드 및 디렉토리 분산 설계

marinRim opened the pull request:
## 🖍️ 이슈번호
- Close #29 

## 🧑‍💻 작업사항

- [x] WebpConverter로 jpg, png를 webp로 변환
- [x] heic, heif 확장자는 jpeg로 변환 후(HeicToJpegConverter) WebpConverter로 넘김
- [x] AWS S3 업로드 시 yyyy/MM/dd/랜덤2글자(영문, 숫자)/파일명 으로 설계
- [x] UserController에서 deleteUserProfiledptj @ModelAttribute -> @ResponseBody로 변경 

## 📸 스크린샷



## 💡 참고사항
- heic, heif 파일은 jpeg 변경해야만 webp로의 변환이 가능하여 `magick`을 로컬에 설치하여 사용함
```java
// ImageMagick 실행
ProcessBuilder processBuilder = new ProcessBuilder(
    "magick",
    tempHeicFile.getAbsolutePath(),
    tempJpegFile.getAbsolutePath()
);
```

- jpg, png, webp 등을 읽는 라이브러리는 build 완료.
- 그러나 webp 쓰기를 하는 라이브러리의 부재로 `cwebp` 로컬에 설치하여 사용함
```java
ProcessBuilder pb = new ProcessBuilder(
    "cwebp",
    "-q", "80",    // (품질 조정 옵션) 80% 퀄리티
    inputFile.getAbsolutePath(),
    "-o",
    outputWebpFile.getAbsolutePath()
);
```

## 🙏 기타사항

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

marinRim committed:
✨ Add converter and Update S3ImageService

marinRim committed:
✨ Update directory structure path by hash and day

marinRim committed:
♻️ Update dealing with image.webp to skip converter

marinRim committed:
✨ Update userprofileImage for webp converting

marinRim committed:
✨ Add to heif converting

marinRim committed:
🩹 Update to ResponseBody

Jgone2 reviewed (approved):
먼저 Converter클래스를 두개로 분리하여 단일 책임 원칙에 준수하며 잘 작성해주신것 같아요! 👍 
Spring에서 HEIC to Webp로 변환하는 과정에서 JPEG로 변환과정을 한번 더 수행해야하는 복잡한 상황 속에서 확장성 있는 구조로 잘 작성해주셨어요! 

리뷰 중에 다른 변경사항 중 개선사항은 필요시 개선해주시면 좋을 것 같고 UserController에서 로그인 상태에서 수행되야하는 요청은 
https://github.com/Grim-eet/grimeet_main/pull/46#discussion_r2070752796
해당 리뷰 참고하여 수정해주신 후 머지 해주시면 감사하겠습니다 👍 

고생하셨어요 :)

Jgone2 commented on src/main/java/com/grimeet/grimeet/common/image/WebpImageConverter.java:
```diff
@@ -0,0 +1,67 @@
+
+        ProcessBuilder pb = new ProcessBuilder(
+                "cwebp",
+                "-q", "80",    // (품질 조정 옵션) 80% 퀄리티
```
해당 설정값에 대해서 상수로 관리하거나 환경변수로 관리하면 추후 서비스 운영측면에 있어서 더 유연하게 운영할 수 있을 것 같아요!

marinRim replied:
피드백 감사합니다! 반영했습니다.

Jgone2 commented on src/main/java/com/grimeet/grimeet/domain/upload/service/S3ImageServiceImpl.java:
```diff
@@ -5,37 +5,53 @@
     @Value("${cloud.aws.s3.bucketName}")
     private String bucketName;
 
+    private static final List<String> ALLOWED_EXTENSIONS = List.of("jpg", "jpeg", "png", "webp", "heic", "heif");
```
해당 리스트와 content_type리스트를 enum으로 처리해도 좋을 것 같아요!

marinRim replied:
타입 안정성이나 확장성을 생각해도 enum 관리가 훨씬 용이하네요. 감사합니다!

Jgone2 commented on src/main/java/com/grimeet/grimeet/domain/upload/service/S3ImageServiceImpl.java:
```diff
@@ -67,40 +87,53 @@ public void deleteImageFromS3(String imageKey) {
+
+        ByteArrayInputStream webpInputStream;
+
+        if ("webp".equalsIgnoreCase(extension)) {
```
확장자별 분기처리를 피룡에 따라 잘 나눠주신 것 같아요! 해당 부분도 추후 enum으로 관리하게 된다면 좋을 것 같아요! 추가적으로 추후에 여러가지 포맷이 추가될 사항을 고려해 switch 문으로 작성하는 것도 좋아보입니다!

marinRim replied:
고민했던 지점입니다. 포맷 추가는 고려하지 않아 if문으로 두었는데 확장성이 확실히 떨어지네요ㅠㅠ 피드백 반영했습니다!

Jgone2 commented on src/main/java/com/grimeet/grimeet/domain/user/controller/UserController.java:
```diff
@@ -101,7 +101,7 @@ public ResponseEntity<UserResponseDto> updateUserProfileImage(@Valid @ModelAttri
     })
     @DeleteMapping("/update/profile-image")
-    public ResponseEntity<UserResponseDto> deleteUserProfileImage(@Valid @ModelAttribute UserDeleteProfileImageRequestDto requestDto) {
+    public ResponseEntity<UserResponseDto> deleteUserProfileImage(@Valid @RequestBody UserDeleteProfileImageRequestDto requestDto) {
```
현재 UserController에서 AuthenticationPrincipal로 user정보를 가져와야할 것 같아요! 이렇게 되면 UpdateProfileImageRequestDto와 UserDeleteProfileImageRequestDto 등 불필요한 Dto를 줄일 수 있을 것 같고, 
비로그인 상태에서 요청시 요청을 할 수 없도록 처리가 가능하기 때문에 안정적인 서비스 운용측면에서도 좋을 것 같아요!

marinRim replied:
피드백 감사합니다!

marinRim committed:
♻️ Delete MailController and Unused tag

marinRim committed:
♻️ Switch commands to constants

marinRim committed:
♻️ Switch List to enum

marinRim committed:
♻️ Swtich if statements to switch-case

marinRim committed:
♻️ Use authenticationPrinciple for current user

UserService now take userId(Long) and find user on DB. But If Controller gives user instance to Service, it will be fast an

marinRim committed:
♻️ Update deleteUserProfileImage with email

marinRim committed:
♻️ Delete user on update parameter

marinRim committed:
♻️ Add authenticationPrincipal on user controller
